File: src/NodeParser/TypeLiteralNodeParser.ts
import ts, { MethodSignature, PropertySignature } from "typescript";
import { Context, NodeParser } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { BaseType } from "../Type/BaseType.js";
import { NeverType } from "../Type/NeverType.js";
import { ObjectProperty, ObjectType } from "../Type/ObjectType.js";
import { ReferenceType } from "../Type/ReferenceType.js";
import { isNodeHidden } from "../Utils/isHidden.js";
import { getKey } from "../Utils/nodeKey.js";

export class TypeLiteralNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
        protected readonly additionalProperties: boolean
    ) {}

    public supportsNode(node: ts.TypeLiteralNode): boolean {
        return node.kind === ts.SyntaxKind.TypeLiteral;
    }
    public createType(node: ts.TypeLiteralNode, context: Context, reference?: ReferenceType): BaseType {
        const id = this.getTypeId(node, context);
        if (reference) {
            reference.setId(id);
            reference.setName(id);
        }

        const properties = this.getProperties(node, context);

        if (properties === undefined) {
            return new NeverType();
        }

        return new ObjectType(id, [], properties, this.getAdditionalProperties(node, context));
    }

    protected getProperties(node: ts.TypeLiteralNode, context: Context): ObjectProperty[] | undefined {
        let hasRequiredNever = false;

        const properties = node.members
            .filter(
                (element): element is PropertySignature | MethodSignature =>
                    ts.isPropertySignature(element) || ts.isMethodSignature(element)
            )
            .filter((propertyNode) => !isNodeHidden(propertyNode))
            .map(
                (propertyNode) =>
                    new ObjectProperty(
                        this.getPropertyName(propertyNode.name),
                        this.childNodeParser.createType(propertyNode.type!, context),
                        !propertyNode.questionToken
                    )
            )
            .filter((prop) => {
                const type = prop.getType();
                if (prop.isRequired() && type instanceof NeverType) {
                    hasRequiredNever = true;
                }
                return !(type instanceof NeverType);
            });

        if (hasRequiredNever) {
            return undefined;
        }

        return properties;
    }

    protected getAdditionalProperties(node: ts.TypeLiteralNode, context: Context): BaseType | boolean {
        const indexSignature = node.members.find(ts.isIndexSignatureDeclaration);
        if (!indexSignature) {
            return this.additionalProperties;
        }

        return this.childNodeParser.createType(indexSignature.type, context) ?? this.additionalProperties;
    }

    protected getTypeId(node: ts.Node, context: Context): string {
        return `structure-${getKey(node, context)}`;
    }

    protected getPropertyName(propertyName: ts.PropertyName): string {
        if (propertyName.kind === ts.SyntaxKind.ComputedPropertyName) {
            const symbol = this.typeChecker.getSymbolAtLocation(propertyName);

            if (symbol) {
                return symbol.getName();
            }
        }

        try {
            return propertyName.getText();
        } catch {
            // When propertyName was programmatically created, it doesn't have a source file.
            // Then, getText() will throw an error. But, for programmatically created nodes,`
            // `escapedText` is available.
            return (propertyName as ts.Identifier).escapedText as string;
        }
    }
}

File: internal/parser/type_literal_node_parser.go
package parser

// Port of src/NodeParser/TypeLiteralNodeParser.ts.

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/tsutils"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

type TypeLiteralNodeParser struct {
	typeChecker          *checker.Checker
	childNodeParser      NodeParser
	additionalProperties bool
}

func NewTypeLiteralNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser, additionalProperties bool) *TypeLiteralNodeParser {
	return &TypeLiteralNodeParser{
		typeChecker:          typeChecker,
		childNodeParser:      childNodeParser,
		additionalProperties: additionalProperties,
	}
}

func (p *TypeLiteralNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindTypeLiteral
}

func (p *TypeLiteralNodeParser) CreateType(node *ast.Node, context *Context, reference *types.ReferenceType) types.Type {
	id := p.getTypeId(node, context)
	if reference != nil {
		reference.SetID(id)
		reference.SetName(id)
	}

	properties, hasRequiredNever := p.getProperties(node, context)
	if hasRequiredNever {
		return &types.NeverType{}
	}

	return types.NewObjectType(id, nil, properties, p.getAdditionalProperties(node, context), false)
}

func (p *TypeLiteralNodeParser) getProperties(node *ast.Node, context *Context) ([]*types.ObjectProperty, bool) {
	hasRequiredNever := false
	var properties []*types.ObjectProperty

	for _, member := range node.Members() {
		if !ast.IsPropertySignatureDeclaration(member) && !ast.IsMethodSignatureDeclaration(member) {
			continue
		}
		if tsutils.IsNodeHidden(member) {
			continue
		}

		property := types.NewObjectProperty(
			p.getPropertyName(member.Name()),
			p.childNodeParser.CreateType(member.Type(), context, nil),
			member.QuestionToken() == nil,
		)

		if types.IsNeverLike(property.Type) {
			if property.Required {
				hasRequiredNever = true
			}
			continue
		}
		properties = append(properties, property)
	}

	if hasRequiredNever {
		return nil, true
	}
	return properties, false
}

func (p *TypeLiteralNodeParser) getAdditionalProperties(node *ast.Node, context *Context) any {
	for _, member := range node.Members() {
		if ast.IsIndexSignatureDeclaration(member) {
			if t := p.childNodeParser.CreateType(member.Type(), context, nil); t != nil {
				return t
			}
			return p.additionalProperties
		}
	}
	return p.additionalProperties
}

func (p *TypeLiteralNodeParser) getTypeId(node *ast.Node, context *Context) string {
	return "structure-" + GetNodeKey(node, context)
}

func (p *TypeLiteralNodeParser) getPropertyName(propertyName *ast.Node) string {
	if propertyName.Kind == ast.KindComputedPropertyName {
		if symbol := tsutils.GetSymbolAtLocation(p.typeChecker, propertyName); symbol != nil {
			return symbol.Name
		}
	}
	// nodeText falls back to the text property for synthesized nodes,
	// mirroring the try/catch with escapedText/text upstream.
	return nodeText(propertyName)
}
